File: src/board.ts
import { sign as ed25519Sign, verify as ed25519Verify } from "node:crypto";
import { assertBlake3SelfTest, blake3HexUtf8 } from "./blake3.ts";
import {
  qualifyFortune5,
  type Fortune5Qualification,
  type Fortune5Requirement,
  type MetricObservation,
  type MetricValue,
  type QualificationContext,
  type Standing,
} from "./fortune5.ts";
import { FORTUNE5_REQUIREMENTS } from "./fortune5.generated.ts";

const DIGEST_RE = /^[0-9a-f]{64}$/;

function canonicalJson(value: unknown): string {
  if (value === null || typeof value !== "object") {
    if (typeof value === "number" && !Number.isFinite(value)) throw new Error("REFUSED:NON_FINITE_CANONICAL_VALUE");
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) return `[${value.map(canonicalJson).join(",")}]`;
  const record = value as Record<string, unknown>;
  return `{${Object.keys(record).sort().map((key) => `${JSON.stringify(key)}:${canonicalJson(record[key])}`).join(",")}}`;
}

export type SignatureAlgorithm = "Ed25519";

export interface ReceiptCoreV2 {
  version: "CASTLE-RECEIPT-V2";
  algorithm: "BLAKE3-256";
  signatureAlgorithm: SignatureAlgorithm;
  payloadDigest: string;
  subject: string;
  metric: string;
  policyDigest: string;
  authorityDigest: string;
  parentDigests: readonly string[];
  keyId: string;
  trustEpoch: number;
  issuedAt: string;
  assuranceDomain: string;
}

export interface ReceiptV2 extends ReceiptCoreV2 {
  receiptDigest: string;
  signature: string;
}

export interface TrustKey {
  keyId: string;
  publicKeyPem: string;
  validFromEpoch: number;
  revokedAtEpoch?: number;
  assuranceDomain: string;
}

export interface TrustStore {
  currentEpoch: number;
  keys: ReadonlyMap<string, TrustKey>;
}

export interface EvidenceInput {
  metric: string;
  value: MetricValue;
  subject: string;
  observedAt: string;
  epistemicClass: "OBSERVED" | "REPLAYED" | "INFERRED";
}

export interface EvidenceBundle {
  observation: MetricObservation;
  receipt: ReceiptV2;
}

export interface ReceiptIssueContext {
  policyDigest: string;
  authorityDigest: string;
  parentDigests?: readonly string[];
  keyId: string;
  trustEpoch: number;
  assuranceDomain: string;
}

function observationPayload(input: EvidenceInput | MetricObservation): EvidenceInput {
  return {
    metric: input.metric,
    value: input.value,
    subject: input.subject,
    observedAt: input.observedAt,
    epistemicClass: input.epistemicClass,
  };
}

function validateDigest(label: string, digest: string): void {
  if (!DIGEST_RE.test(digest)) throw new Error(`REFUSED:INVALID_${label}_DIGEST`);
}

function receiptCoreDigest(core: ReceiptCoreV2): string {
  return blake3HexUtf8(canonicalJson(core));
}

export function issueEvidenceReceipt(
  input: EvidenceInput,
  context: ReceiptIssueContext,
  privateKeyPem: string,
): EvidenceBundle {
  assertBlake3SelfTest();
  validateDigest("POLICY", context.policyDigest);
  validateDigest("AUTHORITY", context.authorityDigest);
  if (!context.keyId || !context.assuranceDomain) throw new Error("REFUSED:INCOMPLETE_RECEIPT_AUTHORITY");
  if (!Number.isInteger(context.trustEpoch) || context.trustEpoch < 0) throw new Error("REFUSED:INVALID_TRUST_EPOCH");
  if (!Number.isFinite(Date.parse(input.observedAt))) throw new Error("REFUSED:INVALID_EVIDENCE_TIME");

  const payloadDigest = blake3HexUtf8(canonicalJson(observationPayload(input)));
  const core: ReceiptCoreV2 = {
    version: "CASTLE-RECEIPT-V2",
    algorithm: "BLAKE3-256",
    signatureAlgorithm: "Ed25519",
    payloadDigest,
    subject: input.subject,
    metric: input.metric,
    policyDigest: context.policyDigest,
    authorityDigest: context.authorityDigest,
    parentDigests: [...new Set(context.parentDigests ?? [])].sort(),
    keyId: context.keyId,
    trustEpoch: context.trustEpoch,
    issuedAt: input.observedAt,
    assuranceDomain: context.assuranceDomain,
  };
  for (const parent of core.parentDigests) validateDigest("PARENT", parent);
  const receiptDigest = receiptCoreDigest(core);
  const signature = ed25519Sign(null, Buffer.from(receiptDigest, "hex"), privateKeyPem).toString("base64");
  const receipt: ReceiptV2 = { ...core, receiptDigest, signature };
  return {
    observation: { ...input, receiptDigest },
    receipt,
  };
}

export interface ReceiptVerification {
  standing: "ALIVE" | "REFUSED";
  reasons: readonly string[];
  verifiedDigests: readonly string[];
}

function verifyReceiptNode(
  receipt: ReceiptV2,
  trust: TrustStore,
  store: ReadonlyMap<string, ReceiptV2>,
  visiting: Set<string>,
  verified: Set<string>,
  reasons: string[],
): void {
  if (verified.has(receipt.receiptDigest)) return;
  if (visiting.has(receipt.receiptDigest)) {
    reasons.push("REFUSED:RECEIPT_DAG_CYCLE");
    return;
  }
  visiting.add(receipt.receiptDigest);

  if (receipt.version !== "CASTLE-RECEIPT-V2") reasons.push("REFUSED:UNSUPPORTED_RECEIPT_VERSION");
  if (receipt.algorithm !== "BLAKE3-256") reasons.push("REFUSED:UNSUPPORTED_RECEIPT_DIGEST_ALGORITHM");
  if (receipt.signatureAlgorithm !== "Ed25519") reasons.push("REFUSED:UNSUPPORTED_SIGNATURE_ALGORITHM");
  if (!DIGEST_RE.test(receipt.receiptDigest)) reasons.push("REFUSED:INVALID_RECEIPT_DIGEST");
  if (!DIGEST_RE.test(receipt.payloadDigest)) reasons.push("REFUSED:INVALID_PAYLOAD_DIGEST");
  if (!DIGEST_RE.test(receipt.policyDigest)) reasons.push("REFUSED:INVALID_POLICY_DIGEST");
  if (!DIGEST_RE.test(receipt.authorityDigest)) reasons.push("REFUSED:INVALID_AUTHORITY_DIGEST");
  if (!Number.isInteger(receipt.trustEpoch) || receipt.trustEpoch < 0 || receipt.trustEpoch > trust.currentEpoch) {
    reasons.push("REFUSED:INVALID_TRUST_EPOCH");
  }

  const { receiptDigest: _digest, signature: _signature, ...core } = receipt;
  const expectedDigest = receiptCoreDigest(core);
  if (expectedDigest !== receipt.receiptDigest) reasons.push("REFUSED:RECEIPT_CONTENT_MISMATCH");

  const key = trust.keys.get(receipt.keyId);
  if (!key) {
    reasons.push("REFUSED:UNTRUSTED_RECEIPT_KEY");
  } else {
    if (receipt.assuranceDomain !== key.assuranceDomain) reasons.push("REFUSED:ASSURANCE_DOMAIN_MISMATCH");
    if (receipt.trustEpoch < key.validFromEpoch) reasons.push("REFUSED:KEY_NOT_YET_VALID");
    if (key.revokedAtEpoch !== undefined && receipt.trustEpoch >= key.revokedAtEpoch) reasons.push("REFUSED:REVOKED_RECEIPT_KEY");
    try {
      const valid = ed25519Verify(
        null,
        Buffer.from(receipt.receiptDigest, "hex"),
        key.publicKeyPem,
        Buffer.from(receipt.signature, "base64"),
      );
      if (!valid) reasons.push("REFUSED:INVALID_RECEIPT_SIGNATURE");
    } catch {
      reasons.push("REFUSED:INVALID_RECEIPT_SIGNATURE");
    }
  }

  for (const parentDigest of receipt.parentDigests) {
    const parent = store.get(parentDigest);
    if (!parent) {
      reasons.push("REFUSED:ORPHAN_RECEIPT_PARENT");
      continue;
    }
    verifyReceiptNode(parent, trust, store, visiting, verified, reasons);
  }

  visiting.delete(receipt.receiptDigest);
  verified.add(receipt.receiptDigest);
}

export function verifyReceiptDag(
  root: ReceiptV2,
  trust: TrustStore,
  store: ReadonlyMap<string, ReceiptV2> = new Map(),
): ReceiptVerification {
  assertBlake3SelfTest();
  const reasons: string[] = [];
  const verified = new Set<string>();
  verifyReceiptNode(root, trust, store, new Set(), verified, reasons);
  return {
    standing: reasons.length === 0 ? "ALIVE" : "REFUSED",
    reasons: reasons.length === 0 ? ["ALIVE:RECEIPT_DAG_VERIFIED"] : [...new Set(reasons)].sort(),
    verifiedDigests: [...verified].sort(),
  };
}

export interface EvidenceAdmission {
  standing: "ALIVE" | "REFUSED";
  reasons: readonly string[];
  observation?: MetricObservation;
}

export function admitEvidence(
  bundle: EvidenceBundle,
  trust: TrustStore,
  store: ReadonlyMap<string, ReceiptV2> = new Map(),
): EvidenceAdmission {
  const reasons: string[] = [];
  const { observation, receipt } = bundle;
  if (observation.receiptDigest !== receipt.receiptDigest) reasons.push("REFUSED:EVIDENCE_RECEIPT_REFERENCE_MISMATCH");
  if (receipt.subject !== observation.subject) reasons.push("REFUSED:EVIDENCE_SUBJECT_MISMATCH");
  if (receipt.metric !== observation.metric) reasons.push("REFUSED:EVIDENCE_METRIC_MISMATCH");
  if (receipt.issuedAt !== observation.observedAt) reasons.push("REFUSED:EVIDENCE_TIME_BINDING_MISMATCH");
  const payloadDigest = blake3HexUtf8(canonicalJson(observationPayload(observation)));
  if (payloadDigest !== receipt.payloadDigest) reasons.push("REFUSED:EVIDENCE_PAYLOAD_MISMATCH");

  const dag = verifyReceiptDag(receipt, trust, store);
  if (dag.standing !== "ALIVE") reasons.push(...dag.reasons);
  return {
    standing: reasons.length === 0 ? "ALIVE" : "REFUSED",
    reasons: reasons.length === 0 ? ["ALIVE:RECEIPTED_EVIDENCE_ADMITTED"] : [...new Set(reasons)].sort(),
    observation: reasons.length === 0 ? observation : undefined,
  };
}

export interface VerifiedQualification {
  standing: Standing;
  qualification: Fortune5Qualification;
  evidenceRefusals: readonly string[];
}

export function qualifyVerifiedFortune5(
  bundles: readonly EvidenceBundle[],
  context: QualificationContext,
  trust: TrustStore,
  receiptStore: ReadonlyMap<string, ReceiptV2> = new Map(),
  requirements: readonly Fortune5Requirement[] = FORTUNE5_REQUIREMENTS,
): VerifiedQualification {
  const admitted: MetricObservation[] = [];
  const refusals: string[] = [];
  for (const bundle of bundles) {
    const admission = admitEvidence(bundle, trust, receiptStore);
    if (admission.standing === "ALIVE" && admission.observation) admitted.push(admission.observation);
    else refusals.push(...admission.reasons.map((reason) => `${bundle.observation.metric}:${reason}`));
  }
  const qualification = qualifyFortune5(admitted, context, requirements);
  return {
    standing: refusals.length > 0 ? "REFUSED" : qualification.standing,
    qualification,
    evidenceRefusals: [...new Set(refusals)].sort(),
  };
}

export interface BoardAdmissionInput {
  enterprise: { context: QualificationContext; evidence: readonly EvidenceBundle[] };
  castle: { context: QualificationContext; evidence: readonly EvidenceBundle[] };
  trust: TrustStore;
  receiptStore?: ReadonlyMap<string, ReceiptV2>;
  castleAssuranceDomain: string;
  independentAssuranceDomain: string;
}

export interface BoardAdmission {
  standing: Standing;
  enterprise: VerifiedQualification;
  castle: VerifiedQualification;
  reasons: readonly string[];
}

function findEvidenceDomain(evidence: readonly EvidenceBundle[], metric: string): string | undefined {
  return evidence.find((item) => item.observation.metric === metric)?.receipt.assuranceDomain;
}

export function qualifyFortune5Board(input: BoardAdmissionInput): BoardAdmission {
  const store = input.receiptStore ?? new Map<string, ReceiptV2>();
  const enterprise = qualifyVerifiedFortune5(input.enterprise.evidence, input.enterprise.context, input.trust, store);
  const castle = qualifyVerifiedFortune5(input.castle.evidence, input.castle.context, input.trust, store);
  const reasons: string[] = [];
  if (enterprise.standing !== "ALIVE") reasons.push(`REFUSED:ENTERPRISE_NOT_ALIVE:${enterprise.standing}`);
  if (castle.standing !== "ALIVE") reasons.push(`REFUSED:CASTLE_NOT_ALIVE:${castle.standing}`);
  if (!input.castleAssuranceDomain || !input.independentAssuranceDomain || input.castleAssuranceDomain === input.independentAssuranceDomain) {
    reasons.push("REFUSED:ASSURANCE_NOT_INDEPENDENT");
  }
  for (const metric of ["independent_verifier_agreement_bps", "board_package_independent_assurance_passed"]) {
    const enterpriseDomain = findEvidenceDomain(input.enterprise.evidence, metric);
    const castleDomain = findEvidenceDomain(input.castle.evidence, metric);
    if (enterpriseDomain !== input.independentAssuranceDomain || castleDomain !== input.independentAssuranceDomain) {
      reasons.push(`REFUSED:INDEPENDENT_ASSURANCE_EVIDENCE:${metric}`);
    }
  }
  return {
    standing: reasons.length === 0 ? "ALIVE" : "REFUSED",
    enterprise,
    castle,
    reasons: reasons.length === 0 ? ["ALIVE:BOARD_ADMISSION_PROVED"] : reasons,
  };
}

export type FailureMode = "FAIL_CLOSED" | "SAFE_DEGRADE" | "LOCAL_CAPABILITY" | "DEFER";

export interface FailureSemanticsInput {
  mode: FailureMode;
  castleAvailable: boolean;
  localCapabilityVerified: boolean;
  receiptChannelAvailable: boolean;
}

export interface FailureSemanticsDecision {
  standing: "ALIVE" | "REFUSED";
  mayActuate: boolean;
  reason: string;
}

export function admitFailureSemantics(input: FailureSemanticsInput): FailureSemanticsDecision {
  if (input.castleAvailable) {
    return {
      standing: input.receiptChannelAvailable ? "ALIVE" : "REFUSED",
      mayActuate: input.receiptChannelAvailable,
      reason: input.receiptChannelAvailable ? "ALIVE:RECEIPTED_PRIMARY_PATH" : "REFUSED:NO_RECEIPT_CHANNEL",
    };
  }
  if (input.mode === "FAIL_CLOSED") return { standing: "ALIVE", mayActuate: false, reason: "ALIVE:FAIL_CLOSED" };
  if (input.mode === "DEFER") return { standing: "ALIVE", mayActuate: false, reason: "ALIVE:DEFERRED" };
  if (!input.receiptChannelAvailable) return { standing: "REFUSED", mayActuate: false, reason: "REFUSED:UNRECEIPTABLE_DEGRADED_DO" };
  if (input.mode === "LOCAL_CAPABILITY" && !input.localCapabilityVerified) {
    return { standing: "REFUSED", mayActuate: false, reason: "REFUSED:LOCAL_CAPABILITY_NOT_VERIFIED" };
  }
  return {
    standing: "ALIVE",
    mayActuate: true,
    reason: input.mode === "LOCAL_CAPABILITY" ? "ALIVE:RECEIPTED_LOCAL_CAPABILITY" : "ALIVE:RECEIPTED_SAFE_DEGRADE",
  };
}

export type MaterialityDimension = "financial" | "operational" | "customer" | "legal" | "regulatory" | "reputational" | "systemic";

export interface MaterialityPolicy {
  policyDigest: string;
  authorityDigest: string;
  perDimensionThresholdBps: Readonly<Record<MaterialityDimension, number>>;
  aggregateThresholdBps: number;
  escalationWithinMs: number;
}

export interface MaterialityEvent {
  id: string;
  subject: string;
  occurredAtEpochMs: number;
  impactBps: Readonly<Record<MaterialityDimension, number>>;
}

export interface MaterialityAssessment {
  material: boolean;
  scoreBps: number;
  triggeringDimensions: readonly MaterialityDimension[];
  escalateByEpochMs?: number;
  policyDigest: string;
  authorityDigest: string;
}

export function assessMateriality(event: MaterialityEvent, policy: MaterialityPolicy): MaterialityAssessment {
  validateDigest("POLICY", policy.policyDigest);
  validateDigest("AUTHORITY", policy.authorityDigest);
  if (!Number.isInteger(policy.aggregateThresholdBps) || policy.aggregateThresholdBps < 0) throw new Error("REFUSED:INVALID_MATERIALITY_POLICY");
  if (!Number.isInteger(policy.escalationWithinMs) || policy.escalationWithinMs < 0) throw new Error("REFUSED:INVALID_ESCALATION_POLICY");
  const dimensions = Object.keys(event.impactBps) as MaterialityDimension[];
  let scoreBps = 0;
  const triggeringDimensions: MaterialityDimension[] = [];
  for (const dimension of dimensions) {
    const impact = event.impactBps[dimension];
    const threshold = policy.perDimensionThresholdBps[dimension];
    if (!Number.isFinite(impact) || impact < 0 || impact > 10000 || !Number.isFinite(threshold) || threshold < 0 || threshold > 10000) {
      throw new Error("REFUSED:INVALID_MATERIALITY_IMPACT");
    }
    scoreBps += impact;
    if (impact >= threshold) triggeringDimensions.push(dimension);
  }
  scoreBps = Math.min(10000, Math.floor(scoreBps / dimensions.length));
  const material = triggeringDimensions.length > 0 || scoreBps >= policy.aggregateThresholdBps;
  return {
    material,
    scoreBps,
    triggeringDimensions: triggeringDimensions.sort(),
    escalateByEpochMs: material ? event.occurredAtEpochMs + policy.escalationWithinMs : undefined,
    policyDigest: policy.policyDigest,
    authorityDigest: policy.authorityDigest,
  };
}

export interface IcfrSubject {
  subject: string;
  processes: readonly string[];
  materialAccounts: readonly string[];
  affectsFinancialReporting: boolean;
}

export interface IcfrClassification {
  subject: string;
  inScope: boolean;
  reasons: readonly string[];
}

export function classifyIcfrSubject(input: IcfrSubject): IcfrClassification {
  const reasons: string[] = [];
  if (input.affectsFinancialReporting) reasons.push("affects-financial-reporting");
  if (input.materialAccounts.length > 0) reasons.push("material-account-impact");
  if (input.processes.some((process) => ["revenue", "payments", "payroll", "purchasing", "inventory", "general-ledger", "financial-reporting"].includes(process))) {
    reasons.push("key-financial-process");
  }
  return { subject: input.subject, inScope: reasons.length > 0, reasons: reasons.sort() };
}

export interface RoleAssignment {
  principal: string;
  role: string;
}

export interface SodViolation {
  principal: string;
  roles: readonly [string, string];
}

export function detectSegregationOfDutyViolations(
  assignments: readonly RoleAssignment[],
  incompatibleRolePairs: readonly (readonly [string, string])[],
): SodViolation[] {
  const byPrincipal = new Map<string, Set<string>>();
  for (const assignment of assignments) {
    const roles = byPrincipal.get(assignment.principal) ?? new Set<string>();
    roles.add(assignment.role);
    byPrincipal.set(assignment.principal, roles);
  }
  const violations: SodViolation[] = [];
  for (const [principal, roles] of [...byPrincipal].sort(([a], [b]) => a.localeCompare(b))) {
    for (const pair of incompatibleRolePairs) {
      if (roles.has(pair[0]) && roles.has(pair[1])) {
        violations.push({ principal, roles: [pair[0], pair[1]] });
      }
    }
  }
  return violations.sort((a, b) => `${a.principal}|${a.roles.join("|")}`.localeCompare(`${b.principal}|${b.roles.join("|")}`));
}

export interface BoardPackage {
  profile: "CASTLE_FORTUNE5_BOARD_V1";
  enterpriseSubject: string;
  castleSubject: string;
  generatedAt: string;
  enterpriseStanding: Standing;
  castleStanding: Standing;
  materialRefusedSubjects: number;
  riskAppetiteBreaches: number;
  controlCount: number;
  evidenceDigest: string;
}

export function buildBoardPackage(
  admission: BoardAdmission,
  generatedAt: string,
  materialRefusedSubjects: number,
  riskAppetiteBreaches: number,
): BoardPackage {
  if (admission.standing !== "ALIVE") throw new Error("REFUSED:BOARD_PACKAGE_WITHOUT_BOARD_ADMISSION");
  if (!Number.isFinite(Date.parse(generatedAt))) throw new Error("REFUSED:INVALID_BOARD_PACKAGE_TIME");
  if (!Number.isInteger(materialRefusedSubjects) || materialRefusedSubjects < 0) throw new Error("REFUSED:INVALID_MATERIAL_REFUSED_COUNT");
  if (!Number.isInteger(riskAppetiteBreaches) || riskAppetiteBreaches < 0) throw new Error("REFUSED:INVALID_RISK_APPETITE_BREACH_COUNT");
  const evidenceDigest = blake3HexUtf8(canonicalJson({
    enterprise: admission.enterprise.qualification.controls,
    castle: admission.castle.qualification.controls,
  }));
  return {
    profile: "CASTLE_FORTUNE5_BOARD_V1",
    enterpriseSubject: admission.enterprise.qualification.subject,
    castleSubject: admission.castle.qualification.subject,
    generatedAt,
    enterpriseStanding: admission.enterprise.standing,
    castleStanding: admission.castle.standing,
    materialRefusedSubjects,
    riskAppetiteBreaches,
    controlCount: FORTUNE5_REQUIREMENTS.length,
    evidenceDigest,
  };
}

export function boardRequirements(): readonly Fortune5Requirement[] {
  return FORTUNE5_REQUIREMENTS;
}

File: tests/board.rs
//! Coverage for `src/board.rs` — the independent `ReceiptV2` (BLAKE3 + real
//! Ed25519) chain used for Fortune-5 evidence admission and board-package
//! qualification. `board.ts` (the TypeScript predecessor) shipped with zero
//! tests; this file closes that gap. Chicago-style throughout: a real
//! `ed25519-dalek` keypair and the real `blake3` crate, state-based
//! assertions on the actual returned structs — no mocking.

use std::collections::{BTreeMap, HashMap};

use castle::board::*;
use castle::fortune5::{EvidenceEpistemicClass, MetricValue, QualificationContext, Standing};
use castle::fortune5_generated::FORTUNE5_REQUIREMENTS;
use ed25519_dalek::SigningKey;

const SUBJECT: &str = "castle:board-test";

fn digest(seed: &str) -> String {
    blake3::hash(seed.as_bytes()).to_hex().to_string()
}

fn trust_store(key_id: &str, verifying_key: ed25519_dalek::VerifyingKey, assurance_domain: &str) -> TrustStore {
    let mut keys = HashMap::new();
    keys.insert(
        key_id.to_string(),
        TrustKey { key_id: key_id.to_string(), verifying_key, valid_from_epoch: 0, revoked_at_epoch: None, assurance_domain: assurance_domain.to_string() },
    );
    TrustStore { current_epoch: 5, keys }
}

fn issue(metric: &str, value: MetricValue, signing_key: &SigningKey, key_id: &str, assurance_domain: &str, trust_epoch: i64) -> EvidenceBundle {
    let input = EvidenceInput { metric: metric.to_string(), value, subject: SUBJECT.to_string(), observed_at: "2026-08-15T19:59:00.000Z".to_string(), epistemic_class: EvidenceEpistemicClass::Observed };
    let context = ReceiptIssueContext {
        policy_digest: digest("policy"),
        authority_digest: digest("authority"),
        parent_digests: vec![],
        key_id: key_id.to_string(),
        trust_epoch,
        assurance_domain: assurance_domain.to_string(),
    };
    issue_evidence_receipt(&input, &context, signing_key).expect("real Ed25519 receipt issuance succeeds")
}

#[test]
fn issued_receipt_verifies_alive_against_its_real_trust_store() {
    let signing_key = SigningKey::from_bytes(&[3u8; 32]);
    let verifying_key = signing_key.verifying_key();
    let trust = trust_store("board-root", verifying_key, "castle-assurance");
    let bundle = issue("deny_by_default", MetricValue::Bool(true), &signing_key, "board-root", "castle-assurance", 5);

    let verification = verify_receipt_dag(&bundle.receipt, &trust, &HashMap::new());
    assert_eq!(verification.standing, Standing::Alive);
    assert_eq!(verification.reasons, vec!["ALIVE:RECEIPT_DAG_VERIFIED".to_string()]);
    assert_eq!(verification.verified_digests, vec![bundle.receipt.receipt_digest.clone()]);
}

#[test]
fn evidence_from_an_untrusted_key_is_refused_not_silently_accepted() {
    let issuing_key = SigningKey::from_bytes(&[9u8; 32]);
    let bundle = issue("deny_by_default", MetricValue::Bool(true), &issuing_key, "board-root", "castle-assurance", 5);

    // Trust store knows a DIFFERENT key under the same key_id (real key-mismatch scenario,
    // not a fabricated signature) — the verifying key here never signed this receipt.
    let other_key = SigningKey::from_bytes(&[10u8; 32]).verifying_key();
    let wrong_trust = trust_store("board-root", other_key, "castle-assurance");

    let verification = verify_receipt_dag(&bundle.receipt, &wrong_trust, &HashMap::new());
    assert_eq!(verification.standing, Standing::Refused);
    assert!(verification.reasons.contains(&"REFUSED:INVALID_RECEIPT_SIGNATURE".to_string()));
}

#[test]
fn admit_evidence_round_trips_a_real_issued_receipt_to_alive() {
    let signing_key = SigningKey::from_bytes(&[4u8; 32]);
    let trust = trust_store("board-root", signing_key.verifying_key(), "castle-assurance");
    let bundle = issue("sbom_coverage_bps", MetricValue::Number(10000.0), &signing_key, "board-root", "castle-assurance", 5);

    let admission = admit_evidence(&bundle, &trust, &HashMap::new());
    assert_eq!(admission.standing, Standing::Alive);
    assert!(admission.observation.is_some());
    assert_eq!(admission.observation.unwrap().metric, "sbom_coverage_bps");
}

#[test]
fn admit_evidence_refuses_when_the_bundles_receipt_digest_reference_is_tampered() {
    let signing_key = SigningKey::from_bytes(&[5u8; 32]);
    let trust = trust_store("board-root", signing_key.verifying_key(), "castle-assurance");
    let mut bundle = issue("sbom_coverage_bps", MetricValue::Number(10000.0), &signing_key, "board-root", "castle-assurance", 5);
    bundle.observation.receipt_digest = digest("a-different-artifact-entirely");

    let admission = admit_evidence(&bundle, &trust, &HashMap::new());
    assert_eq!(admission.standing, Standing::Refused);
    assert!(admission.reasons.contains(&"REFUSED:EVIDENCE_RECEIPT_REFERENCE_MISMATCH".to_string()));
    assert!(admission.observation.is_none());
}

#[test]
fn qualify_verified_fortune5_qualifies_alive_only_from_admitted_receipted_evidence() {
    let signing_key = SigningKey::from_bytes(&[6u8; 32]);
    let trust = trust_store("board-root", signing_key.verifying_key(), "castle-assurance");

    let requirement = FORTUNE5_REQUIREMENTS.iter().find(|r| r.metric == "deny_by_default").unwrap();
    let bundle = issue(requirement.metric, MetricValue::Bool(true), &signing_key, "board-root", "castle-assurance", 5);

    let context = QualificationContext { subject: SUBJECT.to_string(), now_epoch_ms: None, max_evidence_age_ms: None };
    let result = qualify_verified_fortune5(&[bundle], &context, &trust, &HashMap::new(), std::slice::from_ref(requirement));

    assert_eq!(result.standing, Standing::Alive);
    assert_eq!(result.qualification.alive, 1);
    assert!(result.evidence_refusals.is_empty());
}

#[test]
fn qualify_fortune5_board_requires_genuinely_independent_assurance_domains() {
    let signing_key = SigningKey::from_bytes(&[8u8; 32]);
    let trust = trust_store("board-root", signing_key.verifying_key(), "castle-assurance");
    let bundle = issue("deny_by_default", MetricValue::Bool(true), &signing_key, "board-root", "castle-assurance", 5);
    let context = QualificationContext { subject: SUBJECT.to_string(), now_epoch_ms: None, max_evidence_age_ms: None };

    // Same assurance domain on both sides -> not independent -> REFUSED, even though the
    // underlying evidence itself verifies fine.
    let admission = qualify_fortune5_board(BoardAdmissionInput {
        enterprise_context: context.clone(),
        enterprise_evidence: vec![bundle.clone()],
        castle_context: context.clone(),
        castle_evidence: vec![bundle],
        trust: &trust,
        receipt_store: HashMap::new(),
        castle_assurance_domain: "castle-assurance".to_string(),
        independent_assurance_domain: "castle-assurance".to_string(),
    });
    assert_eq!(admission.standing, Standing::Refused);
    assert!(admission.reasons.contains(&"REFUSED:ASSURANCE_NOT_INDEPENDENT".to_string()));
}

#[test]
fn failure_semantics_fail_closed_never_actuates_without_castle() {
    let decision = admit_failure_semantics(&FailureSemanticsInput { mode: FailureMode::FailClosed, castle_available: false, local_capability_verified: false, receipt_channel_available: false });
    assert_eq!(decision.standing, Standing::Alive);
    assert!(!decision.may_actuate);
    assert_eq!(decision.reason, "ALIVE:FAIL_CLOSED");
}

#[test]
fn failure_semantics_refuses_a_degraded_do_with_no_receipt_channel() {
    let decision = admit_failure_semantics(&FailureSemanticsInput { mode: FailureMode::SafeDegrade, castle_available: false, local_capability_verified: false, receipt_channel_available: false });
    assert_eq!(decision.standing, Standing::Refused);
    assert!(!decision.may_actuate);
    assert_eq!(decision.reason, "REFUSED:UNRECEIPTABLE_DEGRADED_DO");
}

#[test]
fn materiality_assessment_triggers_and_schedules_escalation_when_a_dimension_crosses_threshold() {
    let policy = MaterialityPolicy {
        policy_digest: digest("materiality-policy"),
        authority_digest: digest("materiality-authority"),
        per_dimension_threshold_bps: BTreeMap::from([(MaterialityDimension::Financial, 500)]),
        aggregate_threshold_bps: 9000,
        escalation_within_ms: 3_600_000,
    };
    let event = MaterialityEvent { id: "evt:1".to_string(), subject: SUBJECT.to_string(), occurred_at_epoch_ms: 1_000, impact_bps: BTreeMap::from([(MaterialityDimension::Financial, 750)]) };

    let assessment = assess_materiality(&event, &policy).expect("valid materiality inputs");
    assert!(assessment.material);
    assert_eq!(assessment.triggering_dimensions, vec![MaterialityDimension::Financial]);
    assert_eq!(assessment.escalate_by_epoch_ms, Some(1_000 + 3_600_000));
}

#[test]
fn icfr_classification_flags_a_key_financial_process_in_scope() {
    let classification = classify_icfr_subject(&IcfrSubject { subject: SUBJECT.to_string(), processes: vec!["payroll".to_string()], material_accounts: vec![], affects_financial_reporting: false });
    assert!(classification.in_scope);
    assert_eq!(classification.reasons, vec!["key-financial-process".to_string()]);
}

#[test]
fn icfr_classification_leaves_an_unrelated_process_out_of_scope() {
    let classification = classify_icfr_subject(&IcfrSubject { subject: SUBJECT.to_string(), processes: vec!["marketing-analytics".to_string()], material_accounts: vec![], affects_financial_reporting: false });
    assert!(!classification.in_scope);
    assert!(classification.reasons.is_empty());
}

#[test]
fn segregation_of_duty_detects_a_real_incompatible_role_pair() {
    let assignments = vec![
        RoleAssignment { principal: "alice".to_string(), role: "payments-initiate".to_string() },
        RoleAssignment { principal: "alice".to_string(), role: "payments-approve".to_string() },
        RoleAssignment { principal: "bob".to_string(), role: "payments-initiate".to_string() },
    ];
    let incompatible = vec![("payments-initiate".to_string(), "payments-approve".to_string())];

    let violations = detect_segregation_of_duty_violations(&assignments, &incompatible);
    assert_eq!(violations.len(), 1);
    assert_eq!(violations[0].principal, "alice");
    assert_eq!(violations[0].roles, ("payments-initiate".to_string(), "payments-approve".to_string()));
}

#[test]
fn board_requirements_exposes_the_full_generated_forty_control_profile() {
    assert_eq!(board_requirements().len(), 40);
}
